### PR TITLE
[Merged by Bors] - fix: quote string in Cache

### PR DESCRIPTION
**Pull request by digama0**
As [reported on Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60lake.20exe.20cache.20get.60.20fails.20when.20username.20contains.20spaces/near/360687803).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Cache/Requests.lean [hunk 1]
```diff
@@ -28,10 +28,10 @@ section Get
 
 /-- Formats the config file for `curl`, containing the list of files to be downloaded -/
 def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
-  let l ← hashMap.foldM (init := []) fun acc _ hash => do
+  hashMap.foldM (init := "") fun acc _ hash => do
     let fileName := hash.asTarGz
-    pure $ (s!"url = {← mkFileURL fileName none}\n-o {IO.CACHEDIR / fileName}") :: acc
-  return "\n".intercalate l
+    pure $ acc ++ s!"url = {← mkFileURL fileName none}\n-o {
+      (IO.CACHEDIR / fileName).toString.quote}\n"
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Cache/Requests.lean:34]: Note that curl says:

> If the parameter contains whitespace (or starts with : or =), the parameter must be enclosed within quotes. Within double quotes, the following escape sequences are available: \\, \", \t, \n, \r and \v. A backslash preceding any other letter is ignored.

So arguably we should implement that specification, unles it happens to coincide with `string.quote`.

[the remark this answers is not in the document]
Comment (digama0) [reply Cache/Requests.lean:34]: It means "quote for use in lean code", but it's a pretty good match for that spec already, and I don't think we have to worry about usernames containing a vertical tab. I think high ascii is also quoted by `quote`, which might be more of a concern, but I would rather just handle this when it arises.

[the remark this answers is not in the document]
Comment (Kha) [reply Cache/Requests.lean:34]: No, it only encodes control characters.
```lean
#eval "C:\\Users\\Günther".quote  -- "\"C:\\\\Users\\\\Günther\""
```

[the remark this answers is not in the document]
Comment (Kha) [reply Cache/Requests.lean:34]: > Maybe it should just be interpreted, and save the cost of compilation.

That's probably a good idea!

Comment (semorrison) [inline Cache/Requests.lean:34]:
```suggestion
    -- Below we use `String.quote`, which is intended for quoting for use in Lean code
    -- this does not exactly match the requirements for quoting for curl:
    -- ```
    -- If the parameter contains whitespace (or starts with : or =),
    --  the parameter must be enclosed within quotes. 
    -- Within double quotes, the following escape sequences are available:
    --  \, ", \t, \n, \r and \v. 
    -- A backslash preceding any other letter is ignored.
    -- ```
    -- If this becomes an issue we can implement the curl spec.
    pure $ acc ++ s!"url = {← mkFileURL fileName none}\n-o {
      (IO.CACHEDIR / fileName).toString.quote}\n"
```
This is a maximally verbose comment noting Eric's concern. I'm happy with a much shorter one, too. :-)

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**